pandas/tseries/frequencies.py

```diff
@@ -172,15 +173,14 @@ def infer_freq(
         inferer = _TimedeltaFrequencyInferer(index)
         return inferer.get_freq()
 
-    if isinstance(index, Index) and not isinstance(index, DatetimeIndex):
+    if not hasattr(index, "dtype"):
+        pass
+    elif isinstance(index.dtype, object):
```

Comment:
I'm sorry, I'm not familiar enough with the code to review these code changes (more comfortable to review only type annotation changes)